compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -3530,12 +3530,30 @@ class Typer extends Namer
                 typr.println(i"adapt to subtype ${tree.tpe} !<:< $pt")
                 //typr.println(TypeComparer.explained(tree.tpe <:< pt))
                 adaptToSubType(wtp)
-          case CompareResult.OKwithGADTUsed if pt.isValueType =>
+          case CompareResult.OKwithGADTUsed
+          if pt.isValueType
+             && !inContext(ctx.fresh.setGadt(EmptyGadtConstraint)) {
+               (tree.tpe.widenExpr frozen_<:< pt)
+                // we overshot; a cast is not needed, after all. (this happens a lot. We should
+                // find out whether we can make GADTused more precise)
```

Comment:
It seems strange that we'd overshoot often. Could it be that [this](https://github.com/lampepfl/dotty/blob/fix-11220-alt/compiler/src/dotty/tools/dotc/core/TypeComparer.scala#L503) line is a culprit? It should be debuggable reasonably easily, by printing type comparison traces for the subtype check with and without GADT constraints, right at this code line. I'll take a look.

Reply:
"this" gives me a 404 😸

Reply:
Ah, the branch was wrong, it's this link: https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/core/TypeComparer.scala#L503

Reply:
Yes, that's a possible source since we do GADT reasoning before falling back to fourthTry. I tried to swap the order of the two clauses but that gives inference errors.